gusto/linear_solvers.py

```diff
@@ -229,6 +234,321 @@ def solve(self):
+        u, rho = TrialFunctions(M)
+        l0 = TrialFunction(Vtrace)
+        dl = TestFunction(Vtrace)
+
+        n = FacetNormal(state.mesh)
+
+        # Get background fields
+        thetabar = state.fields("thetabar")
+        rhobar = state.fields("rhobar")
+        pibar = thermodynamics.pi(state.parameters, rhobar, thetabar)
+        pibar_rho = thermodynamics.pi_rho(state.parameters, rhobar, thetabar)
+        pibar_theta = thermodynamics.pi_theta(state.parameters, rhobar, thetabar)
+
+        # Analytical (approximate) elimination of theta
+        k = state.k             # Upward pointing unit vector
+        theta = -dot(k, u)*dot(k, grad(thetabar))*beta + theta_in
+
+        # Only include theta' (rather than pi') in the vertical
+        # component of the gradient
+
+        # The pi prime term (here, bars are for mean and no bars are
+        # for linear perturbations)
+        pi = pibar_theta*theta + pibar_rho*rho
+
+        # Vertical projection
+        def V(u):
+            return k*inner(u, k)
+
+        # Specify degree for some terms as estimated degree is too large
+        dxp = dx(degree=(self.quadrature_degree))
+        dS_vp = dS_v(degree=(self.quadrature_degree))
+        dS_hp = dS_h(degree=(self.quadrature_degree))
+        ds_vp = ds_v(degree=(self.quadrature_degree))
+        ds_tbp = ds_t(degree=(self.quadrature_degree)) + ds_b(degree=(self.quadrature_degree))
+
+        # Mass matrix for the trace space
+        tM = assemble(dl('+')*l0('+')*(dS_v + dS_h)
+                      + dl*l0*ds_v + dl*l0*ds_tb)
+
+        Lrhobar = Function(Vtrace)
+        Lpibar = Function(Vtrace)
+        rhopi_solver = LinearSolver(tM, solver_parameters={'ksp_type': 'cg',
+                                                           'pc_type': 'bjacobi',
+                                                           'sub_pc_type': 'ilu'},
+                                    options_prefix='rhobarpibar_solver')
+
+        rhobar_avg = Function(Vtrace)
+        pibar_avg = Function(Vtrace)
+
+        def _traceRHS(f):
+            return (dl('+')*avg(f)*(dS_v + dS_h)
+                    + dl*f*ds_v + dl*f*ds_tb)
+
+        assemble(_traceRHS(rhobar), tensor=Lrhobar)
+        assemble(_traceRHS(pibar), tensor=Lpibar)
+
+        # Project averages of coefficients into the trace space
+        rhopi_solver.solve(rhobar_avg, Lrhobar)
+        rhopi_solver.solve(pibar_avg, Lpibar)
+
+        # Add effect of density of water upon theta
+        if self.moisture is not None:
+            water_t = Function(Vtheta).assign(0.0)
+            for water in self.moisture:
+                water_t += self.state.fields(water)
+            theta = theta / (1 + water_t)
+            thetabar = thetabar / (1 + water_t)
+
+        # "broken" u and rho system
+        Aeqn = (inner(w, (state.h_project(u) - u_in))*dx
+                - beta*cp*div(theta*V(w))*pibar*dxp
+                # + beta*cp*dot(theta*V(w), n)*self.pibar_avg('+')*dS_vp
```

Comment:
General question for all: Do we need this term? I originally thought so since the space for `w` is discontinuous, but theta is still continuous in the vertical. We need to be careful about these surface terms when hybridizing.